### ISSUE TITLE
UsdImagingGLHdEngine::Render(RenderParams params) uses glPush/PopAttrib

### ISSUE THREAD

**Issue by meshula** [PixarAnimationStudios/USD#200]
On Windows 10 with the latest nVidia drivers, and with the context I have created with a core profile, glPush/PopAttribs cause glErrors.

The offending lines are:

````
 glPushAttrib(GL_ENABLE_BIT | GL_POLYGON_BIT);
````
and
````
 glPopAttrib(); // GL_ENABLE_BIT | GL_POLYGON_BIT
````
In the best case (i.e., the case where the function exists ;) ) glPushAttrib can cause pipeline flush. So not desirable ever, really.

The use of GL_POLYGON_BIT in this function appears to be necessary because default poly winding might be modified. To element the use of the attrib bit here, I recommend following the convention that render functions, if they must modify state, should reset values they touched to the redbook state, so explicitly set GL_CCW if params.drawMode dictated otherwise.

The rest of the enables I proposed should be handled as follows at the end of the routine.

GL_CULL_FACE - disable (which is the documented default)

GL_SAMPLE_ALPHA_TO_COVERAGE - enable
https://www.khronos.org/registry/OpenGL-Refpages/es1.1/xhtml/glSampleCoverage.xml

GL_PROGRAM_POINT_SIZE - enable
initial state is seemingly undocumented?

glPolygonMode - GL_FILL
https://www.khronos.org/registry/OpenGL-Refpages/gl4/html/glPolygonMode.xhtml

````
diff --git a/pxr/usdImaging/lib/usdImagingGL/hdEngine.cpp b/pxr/usdImaging/lib/usdImagingGL/hdEngine.cpp
index 94a39d3..abc0b9f 100644
--- a/pxr/usdImaging/lib/usdImagingGL/hdEngine.cpp
+++ b/pxr/usdImaging/lib/usdImagingGL/hdEngine.cpp
@@ -675,8 +676,6 @@ UsdImagingGLHdEngine::Render(RenderParams params)
     glBindVertexArray(vao);
     glBindVertexArray(0);

-    glPushAttrib(GL_ENABLE_BIT | GL_POLYGON_BIT);
-
     // hydra orients all geometry during topological processing so that
     // front faces have ccw winding. We disable culling because culling
     // is handled by fragment shader discard.
@@ -733,15 +732,20 @@ UsdImagingGLHdEngine::Render(RenderParams params)

     glBindVertexArray(0);

-    glPopAttrib(); // GL_ENABLE_BIT | GL_POLYGON_BIT
-
     // XXX: We should not delete the VAO on every draw call, but we currently
     // must because it is GL Context state and we do not control the context.
     glDeleteVertexArrays(1, &vao);
+
+    // reset modified state to initial values
+    glDisable(GL_CULL_FACE);
+    glFrontFace(GL_CCW);
+    glEnable(GL_SAMPLE_ALPHA_TO_COVERAGE);
+    glPolygonMode(GL_FRONT_AND_BACK, GL_FILL);
 }

 /*virtual*/

````

**Comment by jtran56** [comment]
Filed as internal issue #145409.

**Comment by sunyab** [comment]
Closing out old issue, this was addressed in 8af7c8827c0 which was part of the 22.08 release.
